src/parsemultipart.jl

```diff
@@ -0,0 +1,100 @@
+const RETURN_BYTES = [0x0d, 0x0a]
+const DASH_BYTE = 0x2d
+
+const FORMDATA_REGEX = r"Content-Disposition: form-data"
+const NAME_REGEX = r" name=\"(.*?)\""
+const FILENAME_REGEX = r" filename=\"(.*?)\""
+const CONTENTTYPE_REGEX = r"Content-Type: (\S*)"
+
+function find_boundary(bytes::AbstractVector{UInt8}, str, dashes; start::Int = 1)
+    l = length(bytes)
+    i = start
+    cons_dash = 0
+    while i+length(str) <= l
+        @inbounds cons_dash = (bytes[i] == DASH_BYTE) ? cons_dash + 1 : 0
+        if cons_dash >= dashes && bytes[(i+1):(i+length(str))] == str
+            return [i-dashes, i+length(str)]
+        end
+        i += 1
+    end
+    nothing
+end
+
+function find_boundaries(bytes::AbstractVector{UInt8}, str, dashes; start = 1)
+    idxs = Vector{Int}[]
+    j = find_boundary(bytes, str, dashes; start = start)
+    while j !== nothing
+        push!(idxs, j)
+        j = find_boundary(bytes, str, dashes; start = j[2]+1)
+    end
+    return idxs
+end
+
+function find_boundaries(bytes::AbstractVector{UInt8}, boundary; start::Int = 1)
+    m =  match(r"^(-*)(.*)$", boundary)
+    m === nothing && return nothing
+    d, str = m[1], m[2]
+    find_boundaries(bytes, unsafe_wrap(Array{UInt8, 1}, String(str)), length(d); start = start)
```

Comment:
This `unsafe_wrap` combined with the `String` construction looks a bit odd. Is it necessary? (I wondered if it required a `GC.@preserve` but it seems the resulting array knows about the string as the data owner so you should be safe there.)

Also, can you replace the regex match with a simple search `findfirst(c->c != '-', boundary)`? It looks `match(r"^(-*)(.*)$", boundary)` will always succeed except when there's a newline in `boundary`. Is that the intended behavior?

Reply:
I added my first test this evening, I'll take a look at these two items next.

Reply:
I pushed a refactor that addressed the use of the regular expression and attempts to follow RFC 2046 and RFC 7578, the code is completely different. I am not finished refactoring though:
* I am looking at the code that parses the "chunk" now.
* I'd like to change the find_boundary naive string parsing to something like Knuth–Morris–Pratt algorithm (or is there a regex that can search byte arrays in Julia?)

Also, I changed the code that writes to an IO stored in the Multipart to creating an IOBuffer around a view of the body sent from the Request. 
https://github.com/JuliaWeb/HTTP.jl/pull/427/commits/d7d54b2767e8d4c6156b8e00dc58190cfb21263b#diff-1be8563c7ec166272d8a3e9201b93db6R105

I point this out because I don't know if that will cause issues in real code once the Request goes away, although my test ran fine. Does that look okay?